File: worker/src/resolver.ts
import type { CacheStore } from './cache';
import { createCacheStore } from './cache';
import type { ResolveResult, FileInfo, Quality } from './types';

const UA = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36';

function sleep(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}

function atobB64(s: string): string {
  return atob(s.replace(/-/g, '+').replace(/_/g, '/'));
}

function bufToStr(input: string): string {
  return new TextDecoder().decode(
    Uint8Array.from(input, (c) => c.charCodeAt(0)),
  );
}

export function unwrapDl(url: string): string {
  if (!url) return url;
  const m = /link=([^&]+)/.exec(url);
  if (!m) return url;
  let v = m[1];
  try {
    v = decodeURIComponent(v);
  } catch {
    /* keep as-is */
  }
  return v;
}

const QUALITY_RE = /\.(480p|720p|1080p|2160p|4k)\b|\[(480p|720p|1080p|2160p|4k)\]/i;

export function parseQuality(filename: string | null): Quality | null {
  const m = QUALITY_RE.exec(filename || '');
  if (!m) return null;
  return m[1]?.toUpperCase() as Quality ?? m[2]?.toUpperCase() as Quality ?? null;
}

export async function fileInfo(
  directUrl: string,
  cache?: CacheStore,
): Promise<FileInfo> {
  const cacheKey = `fi:${directUrl}`;
  if (cache) {
    const cached = await cache.get(cacheKey);
    if (cached) return JSON.parse(cached);
  }

  const parse = (resp: Response): FileInfo => {
    let length = resp.headers.get('Content-Length');
    const cr = resp.headers.get('Content-Range') || '';
    const mmc = /\/(\d+)\s*$/.exec(cr);
    if (mmc) length = mmc[1];
    const cd = resp.headers.get('Content-Disposition') || '';
    let name = '';
    const m1 = /filename\*=UTF-8''([^;]+)/.exec(cd);
    if (m1) {
      try { name = decodeURIComponent(m1[1]); } catch { name = m1[1]; }
    } else {
      const m2 = /filename="([^"]+)"/.exec(cd);
      if (m2) name = m2[1];
    }
    let size: number | null = null;
    if (length != null && /^\d+$/.test(length)) size = parseInt(length, 10);
    return { size, filename: name };
  };

  const headers = { 'User-Agent': UA };

  try {
    const r = await fetch(directUrl, { method: 'HEAD', headers });
    if (r.ok) {
      const info = parse(r);
      if (cache) await cache.set(cacheKey, JSON.stringify(info), 60_000);
      return info;
    }
  } catch { /* fall through */ }

  try {
    const r = await fetch(directUrl, { headers: { ...headers, Range: 'bytes=0-0' } });
    if (r.ok) {
      const info = parse(r);
      if (cache) await cache.set(cacheKey, JSON.stringify(info), 60_000);
      return info;
    }
  } catch { /* fall through */ }

  return { size: null, filename: '' };
}

// ---------- hubcdn: base64 reurl -> dl wrapper -> bare URL ----------
export async function resolveHubcdn(
  fileUrl: string,
  cache?: CacheStore,
): Promise<ResolveResult> {
  const cacheKey = `hcdn:${fileUrl}`;
  if (cache) {
    const cached = await cache.get(cacheKey);
    if (cached) return JSON.parse(cached);
  }

  let html = '';
  for (let attempt = 0; attempt < 2; attempt++) {
    try {
      const r = await fetch(fileUrl, { headers: { 'User-Agent': UA }, redirect: 'follow' });
      html = await r.text();
      if (/var\s+reurl\s*=\s*"/.test(html)) break;
    } catch { /* retry */ }
    if (attempt === 0) await sleep(700);
  }

  const m = /var\s+reurl\s*=\s*"([^"]+)"/s.exec(html);
  if (!m) return { direct: null, size: null, filename: null, quality: null };
  const reurl = m[1].replace(/\\\//g, '/');

  let direct: string | null = null;
  const mm = /[?&]r=([A-Za-z0-9+/=_-]+)/.exec(reurl);
  if (mm) {
    const b64 = mm[1];
    const pad = (4 - (b64.length % 4)) % 4;
    try {
      direct = bufToStr(atobB64(b64 + '='.repeat(pad)));
    } catch {
      direct = null;
    }
  } else if (reurl.includes('hubcdn.sbs/dl/')) {
    direct = reurl;
  }
  if (direct && !direct.includes('hubcdn.sbs/dl/')) direct = null;
  if (!direct) return { direct: null, size: null, filename: null, quality: null };

  const bare = unwrapDl(direct);
  const info = await fileInfo(bare, cache);
  const result: ResolveResult = {
    direct: bare,
    size: info.size,
    filename: info.filename,
    quality: parseQuality(info.filename),
  };
  if (cache) await cache.set(cacheKey, JSON.stringify(result), 5 * 60_000);
  return result;
}

// ---------- hubdrive: ajax direct-download ----------
export async function resolveHubdrive(
  fileUrl: string,
  cache?: CacheStore,
): Promise<ResolveResult> {
  const cacheKey = `hdrive:${fileUrl}`;
  if (cache) {
    const cached = await cache.get(cacheKey);
    if (cached) return JSON.parse(cached);
  }

  const m = /\/file\/(\d+)/.exec(fileUrl);
  if (!m) return { direct: null, size: null, filename: null, quality: null };
  const fid = m[1];
  const base = fileUrl.slice(0, fileUrl.indexOf('/file/'));

  try {
    await fetch(fileUrl, { headers: { 'User-Agent': UA }, redirect: 'follow' });
  } catch { /* continue anyway */ }

  try {
    const r = await fetch(`${base}/ajax.php?ajax=direct-download`, {
      method: 'POST',
      headers: {
        'User-Agent': UA,
        'X-Requested-With': 'XMLHttpRequest',
        Accept: 'application/json, text/javascript, */*; q=0.01',
        Referer: fileUrl,
        Origin: base,
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body: `id=${encodeURIComponent(fid)}`,
    });
    const data: any = await r.json();
    if (data && data.code === '200' && data.data?.gd) {
      const d = data.data;
      const filename: string = d.n || '';
      let size: number | null = null;
      if (/^\d+$/.test(String(d.s || ''))) size = parseInt(d.s, 10);
      const result: ResolveResult = {
        direct: d.gd,
        size,
        filename,
        quality: parseQuality(filename),
      };
      if (cache) await cache.set(cacheKey, JSON.stringify(result), 5 * 60_000);
      return result;
    }
  } catch { /* fall through */ }

  return { direct: null, size: null, filename: null, quality: null };
}

// ---------- hubcloud: drive -> gamerxyt -> pixeldrain ----------
export async function resolveHubcloud(
  fileUrl: string,
  cache?: CacheStore,
): Promise<ResolveResult> {
  const cacheKey = `hcloud:${fileUrl}`;
  if (cache) {
    const cached = await cache.get(cacheKey);
    if (cached) return JSON.parse(cached);
  }

  let html: string;
  try {
    const r1 = await fetch(fileUrl, { headers: { 'User-Agent': UA }, redirect: 'follow' });
    html = await r1.text();
  } catch {
    return { direct: null, size: null, filename: null, quality: null };
  }

  const m = /href="(https:\/\/gamerxyt\.com\/hubcloud\.php[^"]+)"/.exec(html);
  if (!m) return { direct: null, size: null, filename: null, quality: null };

  let html2: string;
  try {
    const r2 = await fetch(m[1], { headers: { 'User-Agent': UA }, redirect: 'follow' });
    html2 = await r2.text();
  } catch {
    return { direct: null, size: null, filename: null, quality: null };
  }

  const m2 = /var\s+pxl\s*=\s*"(https:\/\/pixeldrain\.[a-z]+\/u\/[A-Za-z0-9]+)"/.exec(html2);
  if (!m2) return { direct: null, size: null, filename: null, quality: null };

  const direct = m2[1].replace('/u/', '/api/file/');
  const info = await fileInfo(direct, cache);
  const result: ResolveResult = {
    direct,
    size: info.size,
    filename: info.filename,
    quality: parseQuality(info.filename),
  };
  if (cache) await cache.set(cacheKey, JSON.stringify(result), 5 * 60_000);
  return result;
}

// ---------- entry: pick the right resolver ----------
export async function resolveDirect(
  fileUrl: string,
  cache?: CacheStore,
): Promise<ResolveResult | null> {
  const c = cache ?? createCacheStore();
  try {
    if (fileUrl.includes('hubdrive.tips')) return await resolveHubdrive(fileUrl, c);
    if (fileUrl.includes('hubcloud.')) return await resolveHubcloud(fileUrl, c);
    if (fileUrl.includes('hubcdn.sbs/file')) return await resolveHubcdn(fileUrl, c);
  } catch {
    return null;
  }
  return null;
}

File: server/app/services/resolver.py
"""Core link resolver – ported from hblinks-server.py to async httpx."""
from __future__ import annotations

import asyncio
import base64
import contextlib
import re
import urllib.parse
from typing import Any, TypedDict, cast

import httpx

from app.services import cache

UA = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36"
HEADERS = {"User-Agent": UA}


class Resolved(TypedDict):
    direct: str | None
    size: int | None
    filename: str | None
    quality: str | None


EMPTY: Resolved = {"direct": None, "size": None, "filename": None, "quality": None}


def unwrap_dl(url: str | None) -> str | None:
    """hubcdn.sbs/dl/?link=<file> → bare <file> URL."""
    if not url:
        return url
    m = re.search(r"link=([^&]+)", url)
    return m.group(1) if m else url


def parse_quality(filename: str) -> str | None:
    m = re.search(
        r"(?<![a-zA-Z0-9])(480p|720p|1080p|2160p|4k)\b", filename, re.I
    ) or re.search(
        r"\[(480p|720p|1080p|2160p|4k)\]", filename, re.I
    )
    return m.group(1).upper() if m else None


def _to_resolved(direct: Any, filename: str, size: int | None) -> Resolved:
    return {
        "direct": direct,
        "size": size,
        "filename": filename,
        "quality": parse_quality(filename),
    }


async def _file_info(client: httpx.AsyncClient, r2_url: str) -> dict[str, Any]:
    """Return {size, filename} for an R2/GDrive direct URL via HEAD then 1-byte range GET."""

    def _parse(headers: httpx.Headers) -> dict[str, Any]:
        length = headers.get("Content-Length")
        cr = headers.get("Content-Range") or ""
        mmc = re.search(r"/(\d+)\s*$", cr)
        if mmc:
            length = mmc.group(1)
        cd = headers.get("Content-Disposition") or ""
        name = ""
        mm = re.search(r"filename\*=UTF-8''([^;]+)", cd)
        if mm:
            name = urllib.parse.unquote(mm.group(1))
        else:
            mm2 = re.search(r'filename="([^"]+)"', cd)
            if mm2:
                name = mm2.group(1)
        size = None
        try:
            size = int(length) if length else None
        except (ValueError, TypeError):
            size = None
        return {"size": size, "filename": name}

    try:
        r = await client.head(r2_url, headers=HEADERS, timeout=15, follow_redirects=True)
        return _parse(r.headers)
    except httpx.HTTPError:
        pass
    try:
        r = await client.get(
            r2_url,
            headers={**HEADERS, "Range": "bytes=0-0"},
            timeout=15,
            follow_redirects=True,
        )
        return _parse(r.headers)
    except httpx.HTTPError:
        return {"size": None, "filename": ""}


def _from_cache(file_url: str) -> Resolved | None:
    cached = cache.get(file_url)
    if cached is None:
        return None
    if isinstance(cached, dict):
        out = cast("Resolved", dict(cached))
        out["direct"] = unwrap_dl(out.get("direct"))
        return out
    # legacy: cache holds a bare dl wrapper string → unwrap it
    return {"direct": unwrap_dl(cached), "size": None, "filename": None, "quality": None}


async def resolve_hubdrive(client: httpx.AsyncClient, file_url: str) -> Resolved:
    """hubdrive.tips/file/<id> → AJAX POST → {direct, size, filename, quality}."""
    m = re.search(r"/file/(\d+)", file_url)
    if not m:
        return EMPTY
    fid = m.group(1)

    # warm up
    with contextlib.suppress(httpx.HTTPError):
        await client.get(file_url, headers=HEADERS, timeout=30)

    try:
        base = file_url[: file_url.index("/file/")]
        headers = {
            "User-Agent": UA,
            "X-Requested-With": "XMLHttpRequest",
            "Accept": "application/json, text/javascript, */*; q=0.01",
            "Referer": file_url,
            "Origin": base,
        }
        body = f"id={urllib.parse.quote(fid)}".encode()
        resp = await client.post(
            f"{base}/ajax.php?ajax=direct-download",
            content=body,
            headers=headers,
            timeout=30,
        )
        data = resp.json()
        if (
            isinstance(data, dict)
            and data.get("code") == "200"
            and isinstance(data.get("data"), dict)
        ):
            d = data["data"]
            if d.get("gd"):
                filename = d.get("n") or ""
                size = None
                try:
                    size = int(d["s"]) if str(d.get("s", "")).isdigit() else None
                except (ValueError, TypeError):
                    size = None
                return _to_resolved(d["gd"], filename, size)
    except (httpx.HTTPError, ValueError):
        pass
    return EMPTY


async def resolve_hubcloud(client: httpx.AsyncClient, file_url: str) -> Resolved:
    """hubcloud.cx/drive/<id> → multi-hop → pixeldrain direct URL."""
    cached = _from_cache(file_url)
    if cached is not None:
        return cached

    try:
        r = await client.get(file_url, headers=HEADERS, timeout=30, follow_redirects=True)
        html = r.text
        m = re.search(r'href="(https://gamerxyt\.com/hubcloud\.php[^"]+)"', html)
        if not m:
            return EMPTY
        r2 = await client.get(m.group(1), headers=HEADERS, timeout=30, follow_redirects=True)
        html2 = r2.text
        m2 = re.search(
            r'var\s+pxl\s*=\s*"(https://pixeldrain\.[a-z]+/u/[A-Za-z0-9]+)"', html2
        )
        if not m2:
            return EMPTY
        direct = re.sub(r"/u/", "/api/file/", m2.group(1), count=1)
        info = await _file_info(client, direct)
        result = _to_resolved(direct, info["filename"], info["size"])
        cache.put(file_url, result)
        return result
    except (httpx.HTTPError, ValueError):
        return EMPTY


async def resolve_hubcdn(client: httpx.AsyncClient, file_url: str) -> Resolved | None:
    """hubcdn.sbs base64 reurl decode → direct link → file_info."""
    cached = _from_cache(file_url)
    if cached is not None:
        return cached

    html = ""
    for attempt in range(2):
        try:
            r = await client.get(file_url, headers=HEADERS, timeout=40, follow_redirects=True)
            html = r.text
            if re.search(r'var\s+reurl\s*=\s*"', html, re.S):
                break
        except httpx.HTTPError:
            pass
        if attempt == 0:
            await asyncio.sleep(0.7)

    if not re.search(r'var\s+reurl\s*=\s*"', html, re.S):
        return None

    direct = None
    m = re.search(r'var\s+reurl\s*=\s*"([^"]+)"', html, re.S)
    if m:
        reurl = m.group(1).replace("\\/", "/")
        mm = re.search(r"[?&]r=([A-Za-z0-9+/=_\-]+)", reurl)
        if mm:
            b64 = mm.group(1)
            padded = b64 + "=" * (-len(b64) % 4)
            try:
                direct = base64.urlsafe_b64decode(padded).decode("utf-8", "replace")
            except Exception:
                direct = None
        elif "hubcdn.sbs/dl/" in reurl:
            direct = reurl

    if direct and "hubcdn.sbs/dl/" not in direct:
        direct = None

    result: Resolved = {
        "direct": None,
        "size": None,
        "filename": None,
        "quality": None,
    }
    if direct:
        bare = unwrap_dl(direct)
        info: dict[str, Any] = {"size": None, "filename": ""}
        if bare:
            info = await _file_info(client, bare)
        result["direct"] = bare
        result["size"] = info["size"]
        result["filename"] = info["filename"]
        result["quality"] = parse_quality(cast("str", info["filename"]))

    if direct:
        cache.put(file_url, result)
    return result


async def resolve_direct(file_url: str) -> Resolved | None:
    """Top-level resolver: dispatches to hubdrive/hubcloud/hubcdn as appropriate."""
    async with httpx.AsyncClient(follow_redirects=True) as client:
        if "hubdrive.tips" in file_url:
            return await resolve_hubdrive(client, file_url)
        if "hubcloud." in file_url:
            return await resolve_hubcloud(client, file_url)
        return await resolve_hubcdn(client, file_url)
